src/envelope.rs

```diff
@@ -0,0 +1,173 @@
+    /// nonce            | ciphertext       | hmac
+    /// nonce_size bytes  | variable length  | hmac_size bytes
+    pub(crate) fn from_bytes(bytes: &[u8]) -> Result<Self, InternalPakeError> {
+        let ciphertext_start = Self::nonce_size();
+        let ciphertext_end = bytes.len() - Self::hmac_size();
+
+        Ok(Self::new(
+            bytes[..ciphertext_start].to_vec(),
+            bytes[ciphertext_start..ciphertext_end].to_vec(),
+            GenericArray::from_slice(&bytes[ciphertext_end..]),
+        ))
+    }
+
+    pub(crate) fn to_bytes(&self) -> Vec<u8> {
+        [&self.nonce[..], &self.ciphertext[..], &self.hmac[..]].concat()
+    }
+
+    pub(crate) fn seal<R: RngCore + CryptoRng>(
+        key: &[u8],
+        plaintext: &[u8],
+        aad: &[u8],
+        rng: &mut R,
+    ) -> Result<(Self, GenericArray<u8, ExportKeySize>), PakeError> {
+        let mut nonce = vec![0u8; Self::nonce_size()];
+        rng.fill_bytes(&mut nonce);
+
+        let h = Hkdf::<Sha256>::new(Some(&nonce), &key);
+        let mut okm = vec![0u8; plaintext.len() + Self::hmac_key_size() + Self::export_key_size()];
+        h.expand(STR_ENVU, &mut okm)
+            .map_err(|_| InternalPakeError::HkdfError)?;
+        let xor_key = &okm[..plaintext.len()];
+        let hmac_key = &okm[plaintext.len()..plaintext.len() + Self::hmac_key_size()];
+        let export_key = &okm[plaintext.len() + Self::hmac_key_size()..];
+
+        let ciphertext: Vec<u8> = xor_key
+            .to_vec()
+            .iter()
+            .zip(plaintext.to_vec().iter())
+            .map(|(&x1, &x2)| x1 ^ x2)
+            .collect();
+
+        let mut hmac =
+            Hmac::<Sha256>::new_varkey(&hmac_key).map_err(|_| InternalPakeError::HmacError)?;
+        hmac.update(&nonce);
+        hmac.update(&ciphertext);
+        hmac.update(&aad);
+
+        Ok((
+            Self::new(nonce, ciphertext.to_vec(), &hmac.finalize().into_bytes()),
+            *GenericArray::from_slice(&export_key),
+        ))
+    }
+
+    pub(crate) fn open(
+        &self,
+        key: &[u8],
+        aad: &[u8],
+    ) -> Result<(Vec<u8>, GenericArray<u8, ExportKeySize>), PakeError> {
+        let h = Hkdf::<Sha256>::new(Some(&self.nonce), &key);
+        let mut okm =
+            vec![0u8; self.ciphertext.len() + Self::hmac_key_size() + Self::export_key_size()];
+        h.expand(STR_ENVU, &mut okm)
+            .map_err(|_| InternalPakeError::HkdfError)?;
+        let xor_key = &okm[..self.ciphertext.len()];
+        let hmac_key = &okm[self.ciphertext.len()..self.ciphertext.len() + Self::hmac_key_size()];
+        let export_key = &okm[self.ciphertext.len() + Self::hmac_key_size()..];
+
+        let mut hmac =
+            Hmac::<Sha256>::new_varkey(&hmac_key).map_err(|_| InternalPakeError::HmacError)?;
+        hmac.update(&self.nonce);
+        hmac.update(&self.ciphertext);
+        hmac.update(aad);
+        if hmac.verify(&self.hmac).is_err() {
```

Comment:
we may want to be explicit —here or in the `thiserror`-powered message for `OpenHmacError`— about the issue revealed by a failure at this step. At a minimum, I think we want to say "opening worked, but the Hmac didnt" rather than what we say now, which is just the later.

Reply:
Opening can no longer fail (because it is an XOR operation instead of a decryption operation now)